evm_arithmetization/src/cpu/kernel/asm/mpt/linked_list/final_tries.asm

```diff
@@ -0,0 +1,231 @@
+    EQ
+    %jumpi(delete_removed_accounts_end)
+    // stack: account_ptr_ptr, root_ptr, storage_ptr_ptr, retdest
+    DUP1
+    %next_account
+    %eq_const(@U256_MAX) // If the next node pointer is @U256_MAX, the node was deleted
+    %jumpi(delete_account)
+    // The account is still there so we need to delete any removed slot.
+    // stack: account_ptr_ptr, root_ptr, storage_ptr_ptr, retdest
+    DUP1
+    MLOAD_GENERAL
+    // stack: key, account_ptr_ptr, root_ptr, storage_ptr_ptr, retdest
+    DUP2
+    %add_const(2)
+    MLOAD_GENERAL // get intitial payload_ptr
+    %add_const(2) // storage_root_ptr_ptr = payload_ptr + 2
+    %mload_trie_data
+    // stack: storage_root_ptr, key, account_ptr_ptr, root_ptr, storage_ptr_ptr, retdest
+    DUP3
+    %increment
+    MLOAD_GENERAL // get dynamic payload_ptr
+    %add_const(2) // storage_root_ptr_ptr = dyn_payload_ptr + 2
+    %stack
+        (storage_root_ptr_ptr, storage_root_ptr, key, account_ptr_ptr, root_ptr, storage_ptr_ptr) ->
+        (key, storage_root_ptr, storage_ptr_ptr, after_delete_removed_slots, storage_root_ptr_ptr, account_ptr_ptr, root_ptr)
+    %jump(delete_removed_slots)
+after_delete_removed_slots:
+    // stack: storage_root_ptr', storage_ptr_ptr', storage_root_ptr_ptr, account_ptr_ptr, root_ptr, retdest
+    SWAP1 SWAP2
+    // stack: storage_root_ptr_ptr, storage_root_ptr', storage_ptr_ptr', account_ptr_ptr, root_ptr, retdest
+    %mstore_trie_data
+    // stack: storage_ptr_ptr', account_ptr_ptr, root_ptr, retdest
+    SWAP1
+    %add_const(4) // The next account in memory
+    // stack: account_ptr_ptr', storage_ptr_ptr', root_ptr, retdest
+    SWAP1 SWAP2 SWAP1
+    %jump(delete_removed_accounts)
+
+delete_removed_accounts_end:
+    // stack: account_ptr_ptr, root_ptr, storage_ptr_ptr, retdest
+    %stack (account_ptr_ptr, root_ptr, storage_ptr_ptr, retdest) -> (retdest, root_ptr)
+    JUMP
+delete_account:
+    // stack: account_ptr_ptr, root_ptr, storage_ptr_ptr, retdest
+    DUP1
+    MLOAD_GENERAL
+    %stack (key, account_ptr_ptr, root_ptr) -> (root_ptr, 64, key, after_mpt_delete, account_ptr_ptr)
+    // Pre stack: node_ptr, num_nibbles, key, retdest
+    // Post stack: updated_node_ptr
+    %jump(mpt_delete)
+after_mpt_delete:
+    // stack: root_ptr', account_ptr_ptr, storage_ptr_ptr, retdest
+    SWAP1
+    %add_const(4)
+    %jump(delete_removed_accounts)
+
+// Delete all slots in `storage_ptr_ptr` with address == `addr` and
+// `storage_ptr_ptr` < `@GLOBAL_METADATA_INITIAL_STORAGE_LINKED_LIST_LEN`.
+// Pre stack: addr, root_ptr, storage_ptr_ptr, retdest
+// Post stack: new_root_ptr, storage_ptr_ptr'.
+delete_removed_slots:
+    // stack: addr, root_ptr, storage_ptr_ptr, retdest
+    DUP3
+    MLOAD_GENERAL
+    // stack: address, addr, root_ptr, storage_ptr_ptr, retdest
+    DUP2
+    EQ
+    %mload_global_metadata(@GLOBAL_METADATA_INITIAL_STORAGE_LINKED_LIST_LEN)
+    DUP5
+    LT
```

Comment:
Don't we want to check whether `storage_ptr_ptr >= list_len`?
If so, we would need to do the `GE` via `LT` on swapped inputs no?

Reply:
I'm not sure I understand: here, we are deleting elements from the initial trie (which have been deleted during the execution). So we want to make sure that the storage slot we are considering for deletion was indeed in the initial trie (and therefore stored at an address < GLOBAL_METADATA_INITIAL_STORAGE_LINKED_LIST_LEN). Or is it not what you meant?

Reply:
Yeah I got confused. Nevermind!